### PR TITLE
Rule: SC2-2-1+SC3-2-5-meta-refresh

### PR DESCRIPTION
**Pull request by jkodu**
# Checklist for Pull Request
🚨Please review the [guidelines for contributing](https://auto-wcag.github.io/auto-wcag/pages/contribute.html) to this repository.

- [x] Make sure you are requesting to **pull a issue/feature/bugfix branch** (right side). Don't request your master branch!
- [x] Make sure you are making a pull request against the **master branch** (left side). Also you should start *your branch* off *master branch*.
- [x] Kindly prefix the pull request name with `DRAFT` or `FINAL` accordingly. For more details, read [Rule Design Guidelines](https://auto-wcag.github.io/auto-wcag/pages/design/review.html)
- [x] Provide a reference to the issue being addressed.

#Reviewing Guidelines
- For `DRAFT` PR's, leave any comments you have and approve if you’ve reviewed the entire pull request. 
- Once there have been three such reviews the pull request can be merged. 
- Any comments left will be taken up in an updated pull request, which will be necessary before a FINAL release”.
- For `FINAL` PR's, review and approve changes to be merged.

# How to Review And Approve
- Go to the “files changed” tab, there you will have the option to leave comments on different lines. 
- Once the review is completed, find the “Review changes” button in the top right, select “approve” and click “Submit review”

### Description
Please describe your pull request.

💔Thank you!

### CODE REVIEW

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 1]
```diff
@@ -1,77 +1,108 @@
...
-- [H76: Using meta refresh to create an instant client-side redirect](http://www.w3.org/TR/WCAG20-TECHS/H76.html)
-- [F40: Failure of Success Criterion 2.2.1 and 2.2.4 due to using meta redirect with a time limit](http://www.w3.org/TR/WCAG20-TECHS/F40.html)
-- [F41: Failure of Success Criterion 2.2.1, 2.2.4, and 3.2.5 due to using meta refresh with a time-out](http://www.w3.org/TR/WCAG20-TECHS/F41.html)
+### Expectation
 
-## Assumptions
+The value before the first semicolon or comma (representing seconds) of the value of the `content` attribute is not a number greater than 0.
 
-- This test assumes no functionality was provided by the website for the user to adjust the timer.
+**Note**: Semi-colon or comma is optional in the `content` attribute. When none is present, the value of the attribute should be considered as a whole. See [meta-refresh](https://www.w3.org/TR/html51/document-metadata.html#statedef-http-equiv-refresh) for parsing instructions.
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: I'd suggest linking to the non-versioned version of the HTML spec: https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh

**Comment by kasperisager** [comment]
Should the limit the applicability to the first valid `<meta http-equiv="refresh">` element encountered, as per the spec?

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md (hunk 1 above)
Comment (kasperisager) [reply _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This is also not quite accurate. Per the spec, space characters can also be used for separating the time and URL of a refresh, so the following is perfectly valid:

```html
<head>
  <meta http-equiv="refresh" content="0 http://example.com" />
</head>
```

This, on the other hand, will cause a redirect after 5 seconds:

```html
<head>
  <meta http-equiv="refresh" content="0: http://example.com" />
  <meta http-equiv="refresh" content="5; http://example.com" />
</head>
```

**Comment by jkodu** [comment]
@annethyme - could you take up the changes suggested by @kasperisager and amend the PR accordingly?

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 2]
```diff
@@ -1,77 +1,165 @@
...
 
-### step1-pass
+The value before the first semicolon, comma or space (representing seconds) of the value of the `content` attribute is not a number greater than 0.
 
-| Property    | Value
-|-------------|----------
-| type        | TestResult
-| outcome     | Passed
-| description |
+**Note**: Semi-colon, comma or space is optional in the `content` attribute. When none is present, the value of the attribute should be considered as a whole. See [meta-refresh](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) for parsing instructions.
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: I'm a bit wary of this note as it's not entirely precise. The value of the attribute should always be considered as a whole as that's the only way to ensure the validity of it. I get what you mean to say by it, but it should be clearer. Changing the expectation to something like the following might alleviate the need for the note altogether:

> The `time` of the `content` attribute, when parsed according to [Refresh state (`http-equiv="refresh"`)](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh), is not greater than 0.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 3]
```diff
@@ -1,77 +1,165 @@
...
+
+```html
+  <head>           
+    <meta http-equiv="refresh" content="foo; URL='https://auto-wcag.github.io/auto-wcag/'" />    
+  </head>  
+```
+
+```html
+<head>
+<meta http-equiv=refresh content=" -00.12 foo">
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This `content` attribute is invalid and therefore inapplicable.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 4]
```diff
@@ -1,77 +1,165 @@
...
+</head>
+```
+
+```html
+<head>
+<meta http-equiv="refresh" content="; 30">
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This `content` attribute is invalid and therefore inapplicable.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 5]
```diff
@@ -1,77 +1,165 @@
...
+</head>
+```
+
+```html
+<head>
+<meta http-equiv="refresh" content="">
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This `content` attribute is invalid and therefore inapplicable.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 6]
```diff
@@ -1,77 +1,165 @@
...
+
+html```
+<head>
+  <meta http-equiv="refresh" content="0 http://example.com" />
+</head>
+```
+
+html```
+<head>
+  <meta http-equiv="refresh" content=" http://example.com" />
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This `content` attribute is invalid and therefore inapplicable.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 7]
```diff
@@ -1,77 +1,165 @@
...
+</head>
+```
+
+html```
+<head>
+  <meta http-equiv="refresh" content="http://example.com" />
```
Comment (kasperisager) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This `content` attribute is invalid and therefore inapplicable.

**Comment by annethyme** [comment]
Requesting review from @WilcoFiers and @jribbens, since you have already contributed your opinions earlier in the process for this rule.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 8]
```diff
@@ -1,77 +1,184 @@
...
-| Property    | Value
-|-------------|----------
-| type        | TestResult
-| outcome     | Failed
-| description | Meta refresh should not be used unless it is instantaneous.
+### Applicability
+
+The rule applies to the first valid `<meta http-equiv="refresh">` element with a `content` attribute in a document.
+
+**Note**: See [meta-refresh](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) for validity of `<meta http-equiv="refresh">` element.
```
Comment (Brynanders) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: Could the note be removed and the word "valid" be linked out to the definition instead?

**Comment by jkodu** [comment]
@annethyme 
Few comments for you to cover...

**Comment by annethyme** [comment]
Update to accommodate Bryn's comment. Please review again.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 9]
```diff
@@ -1,77 +1,182 @@
...
+
+The rule applies to the first [valid](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) `<meta http-equiv="refresh">` element with a `content` attribute in a document.
+
+### Expectation
+
+The `time` of the `content` attribute, when parsed according to [Refresh state (`http-equiv="refresh"`)](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh), is not greater than 0.
+
+## Assumptions  
+
+This test assumes no functionality was provided by the website for the user to adjust the timer.
```
Comment (WilcoFiers) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: It also assumes that the refresh was not essential, which is listed as an exception to this rule in WCAG 2.0.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 10]
```diff
@@ -1,77 +1,182 @@
...
-| type        | TestResult
-| outcome     | Failed
-| description | Meta refresh should not be used unless it is instantaneous.
+### Applicability
+
+The rule applies to the first [valid](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) `<meta http-equiv="refresh">` element with a `content` attribute in a document.
+
+### Expectation
+
+The `time` of the `content` attribute, when parsed according to [Refresh state (`http-equiv="refresh"`)](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh), is not greater than 0.
```
Comment (WilcoFiers) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: I'm wondering if we can't move "when parsed according to [Refresh state (`http-equiv="refresh"`)]" into a note, so we've got an easier to read expectation:

> The `time` of the `content` attribute is not greater than 0.
>
> **note**: See [Refresh state (`http-equiv="refresh"`)](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) for a precise description on how to determine the `time`

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 11]
```diff
@@ -1,77 +1,182 @@
...
+
+## Assumptions  
+
+This test assumes no functionality was provided by the website for the user to adjust the timer. 
+
+## Accessibility Support 
+
+There are no major accessibility support issues known for this rule.
+
+ ## Background
```
Comment (WilcoFiers) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: There shouldn't be a space at the start of this line.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 12]
```diff
@@ -1,77 +1,185 @@
...
+<!-- no content attribute -->
+<head>
+<meta http-equiv="refresh">
+</head>
+```
+
+```html
+<!-- no content attribute -->
+<head>
+<meta http-equiv="">
```
Comment (Brynanders) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: This is duplication and is covered in the previous test case.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 13]
```diff
@@ -1,77 +1,185 @@
...
+
+```html
+<!-- no http-equiv="refresh" attribute -->
+<head>
+<meta content="30">
+</head>
+```
+
+```html
+<!-- no http-equiv="refresh" or content attribute -->
```
Comment (Brynanders) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: If the applicability asks for both, don't we only need a single test case that omits one of the two attributes to cover a scenario where neither are present?

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 14]
```diff
@@ -1,77 +1,185 @@
...
+<head>
+<meta>
+</head>
+```
+
+```html
+<!-- content attribute is invalid and therefore inapplicable. -->
```
Comment (Brynanders) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: Do we need multiple test cases covering 'invalid content attribute'?

Comment (WilcoFiers) [reply _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: We've deleted a number of cases. Issue resolved :)

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 15]
```diff
@@ -1,77 +1,150 @@
...
+* This test assumes no functionality was provided by the website for the user to adjust the timer. 
+* This test assumes that the refresh was not [essential](https://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits-required-behaviors.html#essentialdef), which is listed as a valid exception to SC 2.2.1.
+
+## Accessibility Support 
+
+There are no major accessibility support issues known for this rule.
+
+## Background  
+
+- H76: Using meta refresh to create an instant client-side redirect
```
Comment (jkodu) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: Links?

**Comment by WilcoFiers** [comment]
@carlosapaduarte @ShadowBB can you review this one please?

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 16]
```diff
@@ -1,77 +1,150 @@
...
-| type        | TestResult
-| outcome     | Failed
-| description | Meta refresh should not be used unless it is instantaneous.
+### Applicability
+
+The rule applies to the first [valid](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) `<meta http-equiv="refresh">` element with a `content` attribute in a document.
+
+### Expectation
+
+The `time` of the `content` attribute is not greater than 0.
```
Comment (carlosapaduarte) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: SC 2.2.1 also has a 20 hour exception (in addition to the essential exception already addressed in the assumptions). Will that be handled in another rule of should the expectation be "The time of the content attribute is not between 1 and 72000"?

**Comment by WilcoFiers** [comment]
@carlosapaduarte Updated, good point about the 20 hours. Please approve.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 17]
```diff
@@ -1,77 +1,157 @@
...
+</head>
+```
+
+```html
+<!-- redirects immediately -->
+<head>
+  <meta http-equiv="refresh" content="0 http://example.com" />
+</head>
+```
+
```
Comment (ShadowBB) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: I think the test case above is a duplicate from the first test case. Apart from this minor thing I approve the rule.

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md [hunk 18]
```diff
@@ -1,77 +1,157 @@
...
-| type        | TestResult
-| outcome     | Failed
-| description | Meta refresh should not be used unless it is instantaneous.
+### Applicability
+
+The rule applies to the first [valid](https://www.w3.org/TR/html/document-metadata.html#statedef-http-equiv-refresh) `<meta http-equiv="refresh">` element with a `content` attribute in a document.
+
+### Expectation
+
+The `time` of the `content` attribute is not between 0 and 72000 (20 hours).
```
Comment (carlosapaduarte) [inline _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: I think people might interpret this differently, especially regarding 0 being a valid or invalid result. I would suggest 
The `time`of the `content`attribute is bigger than 0 and smaller than 72000 (20 hours).

Comment (carlosapaduarte) [reply _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: Correction: is 0 or bigger than 72000 (20 hours)

File: _rules/SC2-2-1+SC3-2-5-meta-refresh.md (hunk 17 above)
Comment (WilcoFiers) [reply _rules/SC2-2-1+SC3-2-5-meta-refresh.md]: It's not - it's a broken test case. Removing it. Good catch.

**Comment by WilcoFiers** [comment]
@carlosapaduarte I've updated the rule based on your feedback.

**Comment by annethyme** [comment]
@WilcoFiers and @jkodu: It seems to me that this rule is lost somewhere behind the scenes. It's not a PR or a finished rule on https://auto-wcag.github.io/auto-wcag/pages/rules.html.
Can you guys help me find out what is going on?

**Comment by jkodu** [comment]
@annethyme 
I will have a look shortly.

**Pull request merged**